src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs

```diff
@@ -490,41 +520,52 @@ public override IEnumerable<Cci.INamespaceTypeDefinition> GetTopLevelTypes(EmitC
                 names = null;
             }
 
-            // First time through, we need to push things through NoPiaReferenceIndexer
-            // to make sure we collect all to be embedded NoPia types and members.
-            if (EmbeddedTypesManagerOpt != null && !EmbeddedTypesManagerOpt.IsFrozen)
+            // First time through, we need to push things through TypeReferenceIndexer
+            // to make sure we collect all to be embedded NoPia types and members, as well as detect any usage of NonNullTypes.
+            if ((EmbeddedTypesManagerOpt != null && !EmbeddedTypesManagerOpt.IsFrozen) ||
+                !InjectedSymbolsAreFrozen)
             {
-                noPiaIndexer = new Cci.NoPiaReferenceIndexer(context);
-                Debug.Assert(names != null);
```

Comment:
>Debug.Assert(names != null); [](start = 16, length = 28)

I think we should keep this assert.

Reply:
I remember looking into this, but I don't remember the details. I'll dig up the problematic scenario again.

---
In reply to: [215086893](https://github.com/dotnet/roslyn/pull/29180#discussion_r215086893) [](ancestors = 215086893)

Reply:
Looks like this isn't the problem I thought. Reverted.
Thanks

---
In reply to: [215113343](https://github.com/dotnet/roslyn/pull/29180#discussion_r215113343) [](ancestors = 215113343,215086893)